### ISSUE TITLE
prepend + extend_object still affecting CGI library

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6347]
In #4531 I reported an issue discovered while investigating issues combining `prepend` with `extend_object` in the exotic ways that the CGI library does it. The case there attempted to reproduce the case from CGI as a simple script, but seems to have not been sufficient to exercise the necessary logic; that case began passing (raising the expected error) after #5627 was merged into 9.2.7.0, but the behavior of the CGI library still does not match CRuby.

This has escaped notice (except for a few observant folks like @ahorek) because #4531 was resolved after #5627 appeared to fix the example case and then 3d8847e essentially disabled the native library, so the remaining problem was never addressed.

The specific issue is related to CGI using a combination of `prepend` and `extend_object` to inject native HTML escaping logic into the `CGI::Util` module and `CGI` class, and then using `super` within the native code to fall back on the non-native logic.

This native `super` was the missing link in #4531, and the remaining bug is that the `super` logic used in our native `CGI::EScape` does not properly look up and invoke the super method.

Restoring the pattern from #4531 using the patch below continues to exhibit the same `test_cgi_util.rb` errors reported there:

```diff
diff --git a/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java b/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java
index 7adce40c0e..d92f8ddfb9 100644
--- a/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java
+++ b/core/src/main/java/org/jruby/ext/cgi/escape/CGIEscape.java
@@ -426,10 +426,10 @@ public class CGIEscape implements Library {
     public void load(Ruby runtime, boolean wrap) {
         RubyClass rb_cCGI = runtime.defineClass("CGI", runtime.getObject(), runtime.getObject().getAllocator());
         RubyModule rb_mEscape = rb_cCGI.defineModuleUnder("Escape");
+        RubyModule rb_mUtil = rb_cCGI.defineModuleUnder("Util");
         rb_mEscape.defineAnnotatedMethods(CGIEscape.class);
-        // We do this in cgi/util.rb to work around jruby/jruby#4531.
-//        rb_mUtil.prependModule(rb_mEscape);
-//        rb_mEscape.extend_object(rb_cCGI);
+        rb_mUtil.prependModule(rb_mEscape);
+        rb_mEscape.extend_object(rb_cCGI);
     }
```

I believe this can be fixed by making `Helpers.invokeSuper` properly use the improved lookup logic from #5627.

**Comment by headius** [comment]
Actually it turns out that #4531 was only partially fixed. The case below more accurately reflects what's happening in CGI, and it still errors on JRuby while CRuby prints `:ok`.

```ruby
class Foo
  module X; end
  include X
  extend X
end
module Foo::X
  module Foo::Y
    def blah; super; end # line 8
  end
  prepend Foo::Y
  Foo.extend(Foo::Y)
  def blah; p :ok; end # line 12
end

Foo.blah # line 15
```

When running this with a full backtrace, we see an interesting effect: the `blah` call at line 15 does indeed call the `blah` method at line 8, but the super there re-calls the same method again before failing to super into the `blah` method at line 12:

```
                    callMethodMissing at org/jruby/runtime/Helpers.java:119
                      unresolvedSuper at org/jruby/ir/runtime/IRRuntimeHelpers.java:1204
  invokeSuper2:-unknown-super-target- at blah.rb:8
                                 blah at blah.rb:8
       RUBY$method$blah$4$__VARARGS__ at blah.rb:-1
                                 call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:82
                      unresolvedSuper at org/jruby/ir/runtime/IRRuntimeHelpers.java:1206
  invokeSuper2:-unknown-super-target- at blah.rb:8
                                 blah at blah.rb:8
                                 call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:158
                         cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:354
                                 call at org/jruby/runtime/callsite/CachingCallSite.java:143
                    invokeOther6:blah at blah.rb:15
                               <main> at blah.rb:15
```

The search hierarchy seems to be correct, based on `Foo.ancestors` and `Foo.singleton_class.ancestors`. Printing those produces the following (on both JRuby and CRuby):

```
[Foo, Foo::X, Object, Kernel, BasicObject]
[#<Class:Foo>, Foo::Y, Foo::X, #<Class:Object>, #<Class:BasicObject>, Class, Module, Object, Kernel, BasicObject]
```

It may be that this is not reflecting internal structural differences, since this should not cause `Y.blah` to be called twice (or if it did it should stack overflow as in #4531).

**Comment by headius** [comment]
Analyzing the lookup path seems to show some peculiarities.

I would expect the path in this example to be Foo => Y (extended) => X (extended), where X is actually X' (prepend shim) => Y (prepended) => X. In that sequence, we would expect to see Y#blah called twice.

When stepping through it, that seems to hold up to the point of looking for the superclass of the second prepended Y. It seems to end up as Object, indicating that we've walked off the top of the hierarchy.

There's clearly something missing in JRuby to traverse this hierarchy correctly, or else the hierarchy itself is being structured badly.

**Comment by headius** [comment]
I may have cracked the code.

This logic is the first entry into the actual search for methods on a given class or module:

https://github.com/jruby/jruby/blob/211d935fb1291162287c88016d2134696d128dc7/core/src/main/java/org/jruby/RubyModule.java#L1759-L1771

The logic here applies the current class to the cache entry, which is what we use for supering from that point. The class or wrapper where we found the method is pushed onto the frame, so we know where in context to start walking from to find the next super method.

However the `searchMethodCommon` that it calls also does its own walk of hierarchical structures.

For normal classes, it just looks directly at the method table. But for modules, it walks any hierarchy contained in the module and may find methods from those locations. When it returns to the `searchMethodEntryInner` method, methods from that module subhierarchy will be combined with the host module for the cache entry. This results in the subsequent super logic proceeding from the included module's host hierarchy rather than the module's hierarchy. In this case, it means that the second super proceeds to see the X module's superclass, which is Object's metaclass, and there's no further matches.

I will continue looking into this and try to untangle what the correct hierarchy walk is supposed to do (and also into whether our hierarchy is actually correct).

**Comment by enebo** [comment]
Fixed in PR #6777
